pkg/resources/sku/sku.go

```diff
@@ -0,0 +1,157 @@
+	appsv1 "github.com/openshift/api/apps/v1"
+	appsv12 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	v13 "k8s.io/api/core/v1"
+	"reflect"
+)
+
+type SKU struct {
+	name           string
+	productConfigs map[v1alpha1.ProductName]ProductConfig
+	isUpdated      bool
+}
+
+type ProductConfig struct {
+	productName     v1alpha1.ProductName
+	resourceConfigs map[string]ResourceConfig
+	sku             *SKU
+}
+
+type ResourceConfig struct {
+	Replicas  int32                       `json:"replicas,omitempty"`
+	Resources corev1.ResourceRequirements `json:"resources,omitempty"`
+}
+
+type RateLimit struct {
+	Unit            string  `json:"unit,omitempty"`
+	RequestsPerUnit int64   `json:"requests_per_unit,omitempty"`
+	AlertLimits     []int64 `json:"alert_limits,omitempty"`
+}
+
+type skuConfigReceiver struct {
+	Name      string                    `json:"name,omitempty"`
+	RateLimit RateLimit                 `json:"rate-limiting,omitempty"`
+	Resources map[string]ResourceConfig `json:"resources,omitempty"`
+}
+
+func GetSKU(SKUId string, SKUConfig *corev1.ConfigMap, retSku *SKU) error {
+	received := &[]skuConfigReceiver{}
+	err := json.Unmarshal([]byte(SKUConfig.Data["sku-configs"]), received)
+	if err != nil {
+		return err
+	}
+	skuReceiver := skuConfigReceiver{}
+	for _, sku := range *received {
+		if sku.Name == SKUId {
+			skuReceiver = sku
+			break
+		}
+	}
+	if skuReceiver.Name == "" {
+		return errors.New(fmt.Sprintf("could not find sku config with name '%s'", SKUId))
+	}
+
+	// map of products iterate over that to build the return map
+	products := map[v1alpha1.ProductName][]string{
+		v1alpha1.Product3Scale: {
+			"backend_listener",
+			"backend_worker",
+			"apicast_production",
+		},
+		v1alpha1.ProductRHSSOUser: {
+			"keycloak",
+		},
+		v1alpha1.ProductMarin3r: {
+			"marin3r",
+		},
+	}
+	retSku.name = SKUId
+	retSku.productConfigs = map[v1alpha1.ProductName]ProductConfig{}
+
+	// loop through array of ddcss (deployment deploymentConfig StatefulSets)
+	for product, ddcssNames := range products {
+		pc := ProductConfig{
+			sku:             retSku,
+			productName:     product,
+			resourceConfigs: map[string]ResourceConfig{},
+		}
+		// if any are missing set sane defaults
+		for _, ddcssName := range ddcssNames {
+			pc.resourceConfigs[ddcssName] = skuReceiver.Resources[ddcssName]
+		}
+		retSku.productConfigs[product] = pc
+	}
+	return nil
+}
+
+func (s *SKU) GetProduct(productName v1alpha1.ProductName) ProductConfig {
+	return s.productConfigs[productName]
+}
+
+func (s *SKU) IsUpdated() bool {
+	return s.isUpdated
+}
+
+func (s *SKU) SetUpdated(isUpdated bool) {
+	s.isUpdated = isUpdated
+}
+
+func (p *ProductConfig) GetResourceConfig(ddcssName string) corev1.ResourceRequirements {
+	return p.resourceConfigs[ddcssName].Resources
+}
+
+func (p ProductConfig) GetReplicas(ddcssName string) int32 {
+	return p.resourceConfigs[ddcssName].Replicas
+}
+
+func (p *ProductConfig) Configure(obj interface{}) error {
```

Comment:
This is cool. I'm really happy with how this turned out. Have you tested it is functional?

Reply:
I haven't tested it in cluster but the tests cover the whole function. I started working to update the 3scale one but think I should add an apimanager crd object into the switch.